File: src/main/java/ru/mail/polis/mmishak/KVDaoImpl.java
package ru.mail.polis.mmishak;

import org.jetbrains.annotations.NotNull;
import ru.mail.polis.KVDao;

import javax.xml.bind.DatatypeConverter;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.NoSuchElementException;

public class KVDaoImpl implements KVDao {

    private static final String SEPARATOR = "/";

    private File storageDirectory;

    public KVDaoImpl(File storage) {
        this.storageDirectory = storage;
    }

    @NotNull
    @Override
    public byte[] get(@NotNull byte[] key) throws NoSuchElementException, IOException {
        File file = getFile(key);
        if (!file.exists())
            throw new NoSuchElementException("File with key'" + new String(key) + "' not exits");
        return Files.readAllBytes(file.toPath());
    }

    @Override
    public void upsert(@NotNull byte[] key, @NotNull byte[] value) throws IOException {
        File file = getFile(key);
        if (!file.exists()) {
            final boolean success = file.createNewFile();
            if (!success)
                throw new IOException();
        }
        Files.write(file.toPath(), value);
    }

    @Override
    public void remove(@NotNull byte[] key) throws IOException {
        File file = getFile(key);
        final boolean success = !file.exists() || file.delete();
        if (!success)
            throw new IOException();
    }

    @Override
    public void close() {
        storageDirectory = null;
    }

    @NotNull
    private File getFile(@NotNull byte[] key) {
        return new File(storageDirectory.getAbsolutePath() + SEPARATOR + keyToString(key));
    }

    private String keyToString(byte[] key) {
        return DatatypeConverter.printHexBinary(key);
    }
}

File: src/main/java/ru/mail/polis/mmishak/KVDaoImpl.kt
package ru.mail.polis.mmishak

import org.rocksdb.Options
import org.rocksdb.RocksDB
import org.rocksdb.RocksDBException
import ru.mail.polis.KVDao
import java.io.File
import java.io.IOException
import java.nio.ByteBuffer
import java.util.*
import kotlin.NoSuchElementException

class KVDaoImpl(storageDirectory: File) : KVDao {

    private var database: RocksDB = try {
        RocksDB.open(
            Options().setCreateIfMissing(true),
            storageDirectory.path
        )
    } catch (e: RocksDBException) {
        throw IllegalStateException(e)
    }

    @Throws(NoSuchElementException::class, IOException::class)
    override fun get(key: ByteArray) = try {
        database.get(key) ?: throw NoSuchElementException()
    } catch (e: RocksDBException) {
        throw IOException(e)
    }

    @Throws(IOException::class)
    override fun upsert(key: ByteArray, value: ByteArray) = try {
        database.put(key, value)
    } catch (e: RocksDBException) {
        throw IOException(e)
    }

    @Throws(IOException::class)
    override fun remove(key: ByteArray) = try {
        database.delete(key)
    } catch (e: RocksDBException) {
        throw IOException(e)
    }

    @Throws(IOException::class)
    fun getInternal(key: ByteArray): Data = try {
        val bytes = database.get(key)
        when {
            bytes == null || bytes.isEmpty() -> AbsentData
            bytes[0] == Flag.DELETED.byteValue -> DeletedData(bytes.time)
            else -> ExistsData(bytes.time, bytes.value)
        }
    } catch (e: RocksDBException) {
        throw IOException(e)
    }

    fun upsertInternal(key: ByteArray, value: ByteArray): Boolean = try {
        database.put(key, value.toStoredData(Flag.EXISTS))
        true
    } catch (e: RocksDBException) {
        false
    }

    fun removeInternal(key: ByteArray): Boolean = try {
        database.put(key, byteArrayOf().toStoredData(Flag.DELETED))
        true
    } catch (e: RocksDBException) {
        false
    }

    override fun close() {
        database.close()
    }

    private fun ByteArray.toStoredData(flag: Flag) =
        ByteBuffer.allocate(FLAG_LENGTH + TIME_LENGTH + this.size).apply {
            put(flag.byteValue)
            putLong(System.currentTimeMillis())
            put(this@toStoredData)
        }.array()

    private val ByteArray.time: Long
        get() = Arrays.copyOfRange(this, FLAG_LENGTH, TIME_LENGTH + FLAG_LENGTH).let { timeBytes ->
            ByteBuffer.allocate(TIME_LENGTH).apply {
                put(timeBytes)
                flip()
            }.long
        }

    private val ByteArray.value: ByteArray
        get() = Arrays.copyOfRange(this, FLAG_LENGTH + TIME_LENGTH, this.size)


    companion object {
        init {
            RocksDB.loadLibrary()
        }

        private const val FLAG_LENGTH = 1
        private const val TIME_LENGTH = java.lang.Long.BYTES
    }

    private enum class Flag(val byteValue: Byte) {
        EXISTS(1), DELETED(0);
    }
}


